Use qubit-index newtypes in Rust space

jakelishman opened the pull request:
### Summary

This converts all of our Rust-space components that are concerned with virtual and physical qubits (via the `NLayout` class) to represent those integers with newtypes wrapping them as half-size indices, and changes the `NLayout` API to only allow access via these types and not by raw integers.

The way this is done should have no overhead in time usage from Rust, and is actually a reduction in memory usage because all the qubits are stored at half the width they were previously (for most systems).  This is done to add type safety to all our components that were concerned with the mixing of these two qubits.  The implementation of this commit already turned up the logical bug fixed by gh-10756.

<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Details and comments

The biggest use of `.index()` is in Sabre in the distance matrices.  I considering adding a newtype wrapper around the distance matrix that implements `Index<[PhysicalQubit; 2], Output = f64>` to enforce type-safe access to that during Sabre, but that can potentially be done in a follow-up commit.  I think some other changes I'd potentially like to make around the Sabre interfaces might invalidate that, though, so I've left it for this commit.

I've not made the fix of #10756, so this PR will conflict with that one.  I'll fix up whichever merges second; it's an easy fix.

This PR also takes the opportunity to standardise on `VirtualQubit` throughout the Rust components, which is more consistent with our Python-space usage, and doesn't have the naming clash with the QEC concept of a logical qubit.

*edit*: In (relatively casual) benchmarks I ran of this, I saw no change in runtime of `SabreSwap`, which is as expected.

jakelishman committed:
Use qubit-index newtypes in Rust space

This converts all of our Rust-space components that are concerned with
virtual and physical qubits (via the `NLayout` class) to represent those
integers with newtypes wrapping them as half-size indices, and changes
the `NLayout` API to only allow access via these types and not by raw
integers.

The way this is done should have no overhead in time usage from Rust,
and is actually a reduction in memory usage because all the qubits are
stored at half the width they were previously (for most systems).  This
is done to add type safety to all our components that were concerned
with the mixing of these two qubits.  The implementation of this commit
already turned up the logical bug fixed by gh-10756.

jakelishman commented on qiskit/transpiler/passes/routing/sabre_swap.py:
```diff
@@ -380,7 +380,7 @@ def recurse(dest_dag, source_dag, result, root_logical_map, layout):
                     node.op,
                     [
-                        physical_qubits[layout.logical_to_physical(root_logical_map[q])]
+                        physical_qubits[layout.virtual_to_physical(root_logical_map[q])]
```
Ha, I forgot that I'd called this inner Python-space object `root_logical_map` to match with the Rust-space terminology.  I can replace it now if desired, but I'm about to completely rewrite the Rust/Python interaction during the Sabre reconstruction, so it may well not be worth it.

mtreinish replied:
I think it's fine to leave it as is

mtreinish commented on crates/accelerate/src/nlayout.rs:
```diff
@@ -11,9 +11,76 @@
+            }
+        }
+
+        unsafe impl numpy::Element for $id {
```
Is this the first unsafe block we have in the rust code so far? Maybe it's time we start using `miri` (https://github.com/rust-lang/miri) in CI. This is super minor use now and unlikely to have an issue as we're just implementing the numpy traits which are unsafe because of their nature. But just something to think about moving forward.

jakelishman replied:
I think but am not 100% sure that `unsafe impl` doesn't implicitly allow `unsafe` code within function definitions - it's just a marker that the trait's contract can't be reasonably enforced by the compiler.  In this case, the object has to be safe to be managed by Numpy (give or take, I think that means it needs to be a valid Numpy scalar and not implement `Drop`), but there's no actual unsafe Rust code needed.

mtreinish commented on crates/accelerate/src/nlayout.rs:
```diff
@@ -27,114 +94,125 @@ use hashbrown::HashMap;
-        self.logic_to_phys[logical_bit]
+    /// Get physical bit from virtual bit
+    #[pyo3(text_signature = "(self, virtual, /)")]
+    pub fn virtual_to_physical(&self, r#virtual: VirtualQubit) -> PhysicalQubit {
```
Isn't `r#` a raw string literal? (or at least that's the syntax in rust I know of. Is there some other use case for it here, because this is clearly not a string?

jakelishman replied:
In this case it's a raw identifier rather than a string (no quotes) - the name `virtual` clashes with a reserved but unused keyword in Rust, but I didn't want to expose that detail in the Python-space signature.

mtreinish replied:
Oh, well in that case I guess you could have called it `virt` instead right? It's only ever used positionally here, but this works fine, and I learned new syntax. Although is there any concern if rust ever uses `virtual` for something instead of just reserving it?

jakelishman replied:
I can call it `virt` if you prefer - I don't really mind.  There's no concern if Rust starts using `virtual` - the `r#identifier` syntax is just a lexer construct; it forces it to lex as an identifier token rather than a keyword one and will never clash, and internally, it's treated as an identifier with the same name as the non-raw form:
```rust
fn main() {
    let hello = "world";
    println!("{}", r#hello);
}
```
compiles and prints "world".

mtreinish commented on crates/accelerate/src/nlayout.rs:
```diff
@@ -11,9 +11,76 @@
+            }
+            #[inline]
+            pub fn index(&self) -> usize {
+                self.0 as usize
```
I assume we're not worried about platforms where `usize` is 16 bit :P

jakelishman replied:
Honestly, I didn't actually think about that haha.  Maybe I should have used `from`?

mtreinish replied:
This is fine as is :) we don't support any 16bit platforms and rust only supports one, `msp430-none-elf`, and we're unlikely to ever support building on TI microcontrollers ever as there is no Python or operating system support there.

jakelishman replied:
tbh, I'm pretty impressed that Rust supports _any_ platforms where `usize` is 16-bit.  I'm disappointed that I wrote that cast without even thinking about it, though.

mtreinish commented on crates/accelerate/src/nlayout.rs:
```diff
@@ -27,114 +94,125 @@ use hashbrown::HashMap;
+    pub fn from_virtual_to_physical(virt_to_phys: Vec<PhysicalQubit>) -> PyResult<Self> {
+        let mut phys_to_virt = vec![VirtualQubit(std::u32::MAX); virt_to_phys.len()];
+        for (virt, phys) in virt_to_phys.iter().enumerate() {
+            phys_to_virt[phys.index()] = VirtualQubit(virt.try_into()?);
```
This does seem like quite the edge case on invalid input where `virt_to_phys.len() > u32::MAX` but better to be safe and check that I guess (I also assume clippy would complain about doing an unchecked conversion here).

jakelishman replied:
I don't remember if Clippy complained or anything like that, but I think I tended to err on the side of caution when in Python-callable methods.

mtreinish commented on crates/accelerate/src/sabre_swap/neighbor_table.rs:
```diff
@@ -29,55 +32,75 @@ use rayon::prelude::*;
+                .map(move |v| (NodeIndex::new(u), NodeIndex::new(v.index())))
+        }))
+    }
+}
```
While I agree making this a method of `NeighborTable` makes a lot of sense it seems like an unrelated change here especially given the small details matter for this PR it would be good to keep this to a single logical change just about the `VirtualQubit` and `PhysicalQubit` introduction and use. Can you split this out into a standalone PR?

jakelishman replied:
The answer here's similar to the thing about implementing `Index`: with `.neighbors` now a private field to ensure that the type safety is enforced, I either needed to give `NeighborTable` an iterator method to replace `.neighbors.iter()` or move the coupling-graph constructor here.  I'd chosen this form because it felt like exposing less of the inner API to outside callers, but I'm also ok to split this off into a separate PR either joint with the `impl Index<PhysicalQubit>` (and making `.neighbors` private) PR or a precursor to it.

jakelishman replied:
Reverted in 8018716a5, and the separate PR is #10774.

mtreinish commented on crates/accelerate/src/sabre_swap/neighbor_table.rs:
```diff
@@ -29,55 +32,75 @@ use rayon::prelude::*;
+
+    fn index(&self, index: PhysicalQubit) -> &Self::Output {
+        &self.neighbors[index.index()]
+    }
```
Same here with making it indexable, I think we should do this, but probably in a standalone PR that improves the neighbor table interface in isolation.

jakelishman replied:
The reason I bundled it into this PR is type-safety: if the `NeighborTable.neighbors` field is publically accessible, then it requires that you call `.index()` on the `PhysicalQubit` to get into it, but you could accidentally call `.index()` on a `VirtualQubit` and you'd end up with the same type.  This PR hides the type-unsafe access to `.neighbors` behind an `impl Index<PhysicalQubit>` so it's not possible to accidentally index using a `VirtualQubit`.

That said, I can still split it off into a separate PR if you prefer.

jakelishman replied:
I'd like to do a similar thing with the distance matrix in Sabre as well, but I didn't do that in this PR, which I suppose is another argument in favour of making this change in a separate PR.

jakelishman replied:
Reverted in 8018716a5, and the separate PR is #10774.

mtreinish commented on crates/accelerate/src/sabre_layout.rs:
```diff
@@ -114,15 +114,16 @@ fn layout_trial(
-) -> (NLayout, Vec<usize>, SabreResult) {
-    let num_physical_qubits = distance_matrix.shape()[0];
+) -> (NLayout, Vec<PhysicalQubit>, SabreResult) {
+    let num_physical_qubits: u32 = distance_matrix.shape()[0].try_into().unwrap();
```
This is functionally equivalent to `as u32` right as both will just panic if it's too large for a u32? Is there a difference in the generated code between this vs the straight cast?

jakelishman replied:
`as` casts silently truncate rather than panicking.

mtreinish reviewed (commented):
This looks great, thanks for doing this. Besides shrinking the memory overhead of sabre's rust component and vf2 scoring, leveraging the type checker to validate we're reasoning about the correct type of qubit is a big win. Also I really like the `VirtualQubit::to_phys()` and `PhysicalQubit::to_virt()` syntax it's a lot nicer looking than the syntax before.

Just a couple small comments/questions inline.

jakelishman committed:
Separate out type-safe enforcement in `NeighborTable` to another commit

This reverts the changes to the coupling-map generation and the indexing
into `NeighborTable` that were designed to make it more type safe in
favour of doing that in a separate commit.

mtreinish reviewed (approved):
LGTM, thanks for splitting out the `NeighborTable` changes.
